### PR TITLE
Add 2D Group Convolution Backward Data and Weights update solvers. Simplify and unify 3d group conv tests

### PR DESCRIPTION
**Pull request by amberhassaan**
Follow up to #2536 where I use the generic test fixture class (and macros) for 3D group convolution gtests. 

Copy of description from #2536: 
This PR provides CK-based Backward-Data and Backward-Weight update solvers for 2D Group Convolution. This completes the solver picture for 2D group convolutions (except for f8, which has its own solvers).

Also included in this PR is a generic design for the GTest Fixture class that handles 2D and 3D (Fwd,Bwd,Wrw) passes at the same time, so testing code can reuse this class and related macros. A future PR will use this class for 3D grouped conv gtests (arriving shortly).

### CODE REVIEW

**Review by junliume — APPROVED** [review]

File: src/solver.cpp [hunk 1]
```diff
@@ -607,6 +607,14 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
...
     Register(registry, ++id, Primitive::Batchnorm, batchnorm::BnCKBwdBackward{}.SolverDbId());
     Register(registry, ++id, Primitive::Batchnorm, batchnorm::BnCKFwdTraining{}.SolverDbId());
+    RegisterWithSolver(registry,
+                       ++id,
+                       conv::ConvHipImplicitGemmGroupBwdXdlops{},
+                       miopenConvolutionAlgoImplicitGEMM);
+    RegisterWithSolver(registry,
+                       ++id,
+                       conv::ConvHipImplicitGemmGroupWrwXdlops{},
+                       miopenConvolutionAlgoImplicitGEMM);
```
Comment (CAHEK7) [inline src/solver.cpp]: The order of ids must not be altered for backward compatibility.
Please move it to the end of the list.

File: src/include/miopen/handle.hpp [hunk 2]
```diff
@@ -211,6 +211,15 @@ struct MIOPEN_EXPORT Handle : miopenHandle
         return result;
     }
 
+    template <class V>
+    void ReadToVec(const Allocator::ManageDataPtr& ddata, V& output_vec)
+    {
+        using T = typename V::value_type;
+        assert(ddata);
+        assert(output_vec.size() > 0);
+        this->ReadTo(output_vec.data(), ddata, output_vec.size() * sizeof(T));
```
Comment (CAHEK7) [inline src/include/miopen/handle.hpp:208]: Should we also check that `ddata` has at least `output_vec.size() * sizeof(T)` bytes allocated?

**Review by CAHEK7 — CHANGES_REQUESTED** [review]
Minor `ReadToVec` comment is here.
But the most important part is to keep all old solver ids unaltered.

File: src/solver.cpp (hunk 1 above)
Comment (amberhassaan) [reply src/solver.cpp]: Will do.

File: src/include/miopen/handle.hpp (hunk 2 above)
Comment (amberhassaan) [reply src/include/miopen/handle.hpp:208]: Yes, I wanted that but today we don't store size of the buffer with the ManagedPtr, so I plan on doing this in the near future.

**Review by CAHEK7 — APPROVED** [review]

File: src/solver/conv_hip_implicit_gemm_grouped_bwd_xdlops.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+{
+#if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
+    if(miopen::IsDisabled(ENV(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_GROUP_BWD_XDLOPS)))
+        return false;
+    if(miopen::IsEnabled(ENV(MIOPEN_DEBUG_CONVOLUTION_DETERMINISTIC)))
+        return false;
+    if(problem.GetInDataType() != problem.GetWeightsDataType() ||
+       problem.GetWeightsDataType() != problem.GetOutDataType() ||
+       problem.GetInDataType() != problem.GetOutDataType())
+        return false;
```
Comment (averinevg) [inline src/solver/conv_hip_implicit_gemm_grouped_bwd_xdlops.cpp:308]:
```suggestion
        return false;
    if(problem.HasNonPackedTensors())
        return false;
```

File: src/solver/conv_hip_implicit_gemm_grouped_wrw_xdlops.cpp [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+{
+#if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
+    if(miopen::IsDisabled(ENV(MIOPEN_DEBUG_GROUP_CONV_IMPLICIT_GEMM_HIP_WRW_XDLOPS)))
+        return false;
+    if(miopen::IsEnabled(ENV(MIOPEN_DEBUG_CONVOLUTION_DETERMINISTIC)))
+        return false;
+    if(problem.GetInDataType() != problem.GetWeightsDataType() ||
+       problem.GetWeightsDataType() != problem.GetOutDataType() ||
+       problem.GetInDataType() != problem.GetOutDataType())
+        return false;
```
Comment (averinevg) [inline src/solver/conv_hip_implicit_gemm_grouped_wrw_xdlops.cpp:304]:
```suggestion
        return false;
    if(problem.HasNonPackedTensors())
        return false;
```

File: src/solver/conv_hip_implicit_gemm_grouped_bwd_xdlops.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+    [[maybe_unused]] const ProblemDescription& problem) const
+{
+#if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
+    if(miopen::IsDisabled(ENV(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_GROUP_BWD_XDLOPS)))
+        return false;
+    if(miopen::IsEnabled(ENV(MIOPEN_DEBUG_CONVOLUTION_DETERMINISTIC)))
+        return false;
+    if(problem.GetInDataType() != problem.GetWeightsDataType() ||
+       problem.GetWeightsDataType() != problem.GetOutDataType() ||
+       problem.GetInDataType() != problem.GetOutDataType())
```
Comment (averinevg) [inline src/solver/conv_hip_implicit_gemm_grouped_bwd_xdlops.cpp]:
```suggestion
    if(problem.HasMixedDataTypes())
```

File: src/solver/conv_hip_implicit_gemm_grouped_wrw_xdlops.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    [[maybe_unused]] const ProblemDescription& problem) const
+{
+#if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
+    if(miopen::IsDisabled(ENV(MIOPEN_DEBUG_GROUP_CONV_IMPLICIT_GEMM_HIP_WRW_XDLOPS)))
+        return false;
+    if(miopen::IsEnabled(ENV(MIOPEN_DEBUG_CONVOLUTION_DETERMINISTIC)))
+        return false;
+    if(problem.GetInDataType() != problem.GetWeightsDataType() ||
+       problem.GetWeightsDataType() != problem.GetOutDataType() ||
+       problem.GetInDataType() != problem.GetOutDataType())
```
Comment (averinevg) [inline src/solver/conv_hip_implicit_gemm_grouped_wrw_xdlops.cpp]:
```suggestion
    if(problem.HasMixedDataTypes())
```

**Review by averinevg — CHANGES_REQUESTED** [review]

**Comment by amberhassaan** [comment]
@averinevg : Please approve the PR. All changes applied.

**Review by averinevg — APPROVED** [review]

**Review by iq136boy — APPROVED** [review]

**Pull request merged**